kernel/events.c

```diff
@@ -0,0 +1,188 @@
+}
+
+void z_impl_k_event_post(struct k_event *event, uint32_t events,
+			 unsigned int options)
+{
+	k_spinlock_key_t  key;
+	struct k_thread  *thread;
+	unsigned int      wait_condition;
+	bool              clear = false;
+	struct k_thread  *head = NULL;
+
+	key = k_spin_lock(&event->lock);
+
+	if ((options & K_EVENT_DELIVERY_MASK) == K_EVENT_ACCUMULATE) {
+		events |= event->events;
+	}
+
+	event->events = events;
+
+	/*
+	 * Posting an event has the potential to wake multiple pended threads.
+	 * It is desirable to unpend all affected threads simultaneously. To
+	 * do so, this must be done in three steps as it is unsafe to unpend
+	 * threads from within the _WAIT_Q_FOR_EACH() loop.
+	 *
+	 * 1. Create a linked list of threads to unpend.
+	 * 2. Unpend each of the threads in the linked list
+	 * 3. Ready each of the threads in the linked list
+	 */
+
+	_WAIT_Q_FOR_EACH(&event->wait_q, thread) {
+		wait_condition = thread->event_options & K_EVENT_WAIT_MASK;
+
+		if (!test_for_match(thread->events, events, wait_condition)) {
+			continue;
+		}
+
+		/*
+		 * The wait conditions have been satisfied. Add this thread
+		 * to the list of threads to unpend.
+		 */
+
+		thread->next_link = head;
+		head = thread;
+
+		if (thread->event_options & K_EVENT_POST_CLEAR) {
+			clear = true;
+		}
+	}
+
+	if (head != NULL) {
+		thread = head;
+		do {
+			z_unpend_thread(thread);
+			arch_thread_return_value_set(thread, 0);
+			thread->events = events;
+			thread = thread->next_link;
+		} while (thread != NULL);
+
+		if (clear) {
+			event->events = 0;
+		}
+
+		z_ready_linked_threads(head);
+	}
+
+	z_reschedule(&event->lock, key);
+}
+
+uint32_t z_impl_k_event_wait(struct k_event *event, uint32_t events,
+			     unsigned int options, k_timeout_t timeout)
+{
+	uint32_t  rv = 0;
+	unsigned int  wait_condition;
+	struct k_thread  *thread;
+
+	__ASSERT(((arch_is_in_isr() == false) ||
+		  K_TIMEOUT_EQ(timeout, K_NO_WAIT)), "");
+
+	wait_condition = options & K_EVENT_WAIT_MASK;
+	thread = z_current_get();
+
+	k_spinlock_key_t  key = k_spin_lock(&event->lock);
+
+	if (options & K_EVENT_PRE_CLEAR) {
+		event->events = 0;
+	}
+
+	/*
+	 * Test if the match conditions have already been met. Note that a
+	 * match will be detected if the caller wants no events.
+	 */
+
+	if (test_for_match(events, event->events, wait_condition)) {
+		rv = event->events;
+		k_spin_unlock(&event->lock, key);
+
+		if (options & K_EVENT_POST_CLEAR) {
+			event->events = 0;
+		}
+		goto out;
+	}
+
+	/* Match conditions have not been met. */
+
+	if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
+		k_spin_unlock(&event->lock, key);
+		goto out;
+	}
+
+	/*
+	 * The caller must pend to wait for the match. Save the desired
+	 * set of events in the k_thread structure.
+	 */
+
+	thread->events = events;
+	thread->event_options = options;
+
+	if (z_pend_curr(&event->lock, key, &event->wait_q, timeout) == 0) {
+		/* Retrieve the set of events that woke the thread */
+		rv = thread->events;
+	}
+
+out:
```

Comment:
Almost certainly you want the unlock here, with an early return after z_pend_curr() (which will unlock atomically of course).